src/Core/functions.php

```diff
@@ -325,3 +326,113 @@ function deprecationWarning(string $version, string $message, int $stackFrame =
         trigger_error($message, E_USER_DEPRECATED);
     }
 }
+
+/**
+ * Converts the given value to a string.
+ *
+ * This method attempts to convert the given value to a string.
+ * If the value is already a string, it returns the value as it is.
+ * If the conversion is not possible, it returns NULL.
+ *
+ * @param mixed $value The value to be converted.
+ * @return ?string Returns the string representation of the value, or null if the value is not a string.
+ * @since 5.1.0
+ */
+function toString(mixed $value): ?string
+{
+    if (is_string($value)) {
+        return $value;
+    }
+    if (is_int($value)) {
+        return (string)$value;
+    }
+    if (is_bool($value)) {
+        return $value ? '1' : '0';
+    }
+    if (is_float($value)) {
+        if (is_nan($value) || is_infinite($value)) {
+            return null;
+        }
+        try {
+            $return = json_encode($value, JSON_THROW_ON_ERROR);
+        } catch (JsonException) {
+            $return = null;
+        }
+
+        if ($return === null || str_contains($return, 'e')) {
+            $return = rtrim(sprintf('%.' . (PHP_FLOAT_DIG + 3) . 'F', $value), '.0');
+        }
+
+        return $return;
+    }
+    if ($value instanceof Stringable) {
+        return (string)$value;
+    }
+
+    return null;
+}
+
+/**
+ * Converts a value to an integer.
+ *
+ * This method attempts to convert the given value to an integer.
+ * If the conversion is successful, it returns the value as an integer.
+ * If the conversion fails, it returns NULL.
+ *
+ * String values are trimmed using trim().
+ *
+ * @param mixed $value The value to be converted to an integer.
+ * @return int|null Returns the converted integer value or null if the conversion fails.
+ * @since 5.1.0
+ */
+function toInt(mixed $value): ?int
+{
+    if (is_int($value)) {
+        return $value;
+    }
+    if (is_string($value)) {
+        $value = filter_var($value, FILTER_VALIDATE_INT, FILTER_NULL_ON_FAILURE);
+
+        return $value === PHP_INT_MIN ? null : $value;
+    }
+    if (is_float($value)) {
+        /**
+         * @link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER
+         * 9007199254740991 = 2^53-1 = the maximum safe integer that can be represented without losing precision.
```

Comment:
I'm not sure this same constraint applies to PHP. We should be able to go to 2^64-1 or `PHP_INT_MAX`

Reply:
Yes, this constant also has the same value in PHP. I'm attaching a test that will find the first values that are no longer correctly typecasted.

Test:
```php
public function testMargin(): void
{
    $margin = 9007199254740991;

    $errors = 0;
    for ($i = $margin - 10_000_000; $i <= $margin + 10_000; $i++) {
        if ((int)(float)$i !== $i) {
            ++$errors;
            echo $i . PHP_EOL;
            if ($errors > 20) {
                break;
            }
        }
    }

    $this->markTestSkipped();
}
```

Results:
```
9007199254740993
9007199254740995
9007199254740997
9007199254740999
9007199254741001
9007199254741003
9007199254741005
9007199254741007
9007199254741009
9007199254741011
9007199254741013
9007199254741015
9007199254741017
9007199254741019
9007199254741021
9007199254741023
9007199254741025
9007199254741027
9007199254741029
9007199254741031
9007199254741033
```

Reply:
I found in PHP documentation that it might not always be the same constant. It might be something else on exotic systems.

> Floating point numbers have limited precision. Although it depends on the system, PHP typically uses the IEEE 754 double precision format..
https://www.php.net/manual/en/language.types.float.php

So I changed https://github.com/cakephp/cakephp/pull/17295/commits/af822bbb28fd3289423940201d6cd5160811ed32 the logic to use `(int)$value`, but around `PHP_INT_MIN`/`PHP_INT_MAX`, it returns very strange values.
Beyond 2^53-1, there are type errors like `9007199254740993.0 ==> 9007199254740992`

But I'm okay with this solution. Such large numbers are not used often. And programmers rarely take care of them.